### PR TITLE
Rails 5 update

### PR DESCRIPTION
**Pull request by sonalkr132**
Please read commit message for explanation of changes.

It is very likely that I may have missed deprecations which didn't show up in test, feel free to open issues when you see them.

I have removed [`quite_assets`](https://github.com/evrone/quiet_assets) gem because their rails requirement wouldn't have allowed update and looks like gem is abandoned. The feature is now [implemented in sprocket-rails](https://github.com/rails/sprockets-rails/pull/355), we just have to be wait for version release.

I really want to compare before and after performance, object allocation etc. I would really appreciate it if someone pointed me to relevant resources or took out time for general advice.

### CODE REVIEW

**Comment by dwradcliffe** [comment]
I haven't looked at the commits yet, but if there are any of these things that are compatible with both Rails 5 and the current version we should get those into separate PRs and merge them sooner. Smaller changes are always better to test.

cc: @arthurnn @rafaelfranca

**Comment by sonalkr132** [comment]
> if there are any of these things that are compatible with both Rails 5 and the current version we should get those into separate PRs and merge them sooner.

I am not sure if that would be a good idea. You won't be able to see anything in the logs other than deprecation warnings.

This PR is not as bad it looks. Just gemfiles change amount for 128 lines, `id: ..` => `params: { id: .. }` cloc at 138 lines and adding version to migration is 88 line (82% total change).

**Comment by sonalkr132** [comment]
Thank you for checking out the PR @kmcphillips :octocat:

>  you need to add config.assets.quiet = true to environments/development.rb for the equivalent behaviour.

I didn't add it because I didn't want to add a configuration which did nothing at the point(commit) it was added. `quite_assets` isn't anything critical as in we can live without it for a day or two. I hope it makes sense :sweat_smile:

**Comment by kmcphillips** [comment]
#1352 simplifies this a bit. `sprockets-rails` 3.1.0 was released today.

**Comment by sonalkr132** [comment]
I have great new.. :disappointed: 

It has treading issue and as they go.. it is not reliably reproduceable. Failing [build1](https://travis-ci.org/sonalkr132/rubygems.org/jobs/140843290#L922), [build2](https://travis-ci.org/rubygems/rubygems.org/jobs/140831890#L1126), [build3](https://travis-ci.org/sonalkr132/rubygems.org/jobs/140841700#L925).
Almost all of them are on [`increment` function](https://github.com/rubygems/rubygems.org/blob/master/app/models/gem_download.rb#L45-L52) of GemDownload. How do I even report this?

**Comment by sonalkr132** [comment]
I have updated this to 5.0.0. Treading error was fixed here: https://github.com/rails/rails/pull/25707
I guess we will have to wait for 5.0.1.

File: config/application.rb [hunk 1]
```diff
@@ -28,6 +25,9 @@ class Application < Rails::Application
     config.plugins = [:dynamic_form]
 
     config.autoload_paths << Rails.root.join('lib')
+
+    # Do not halt callback chains when a callback returns false.
+    ActiveSupport.halt_callback_chains_on_return_false = false
   end
```
Comment (prathamesh-sonpatki) [inline config/application.rb]: @sonalkr132 Did you run `rails app:update`, this config will go to `config/initializers/new_framework_defaults.rb` after that.

Comment (sonalkr132) [inline config/application.rb]: I was hoping I would change config after we are sure that changes up to this point don't break anything.
However, now it seems like it is going to be a while before this gets deployed.. I have added the changes of `rails app:update`. Thanks for having a look at the PR :)

File: Gemfile [hunk 2]
```diff
@@ -53,9 +53,7 @@ group :development, :test do
   gem 'toxiproxy', '~> 0.1.3'
 end
 
-group :development do
-  gem 'rails-erd'
-end
+gem 'rails-erd', group: :development
```
Comment (simi) [inline Gemfile]: Is this needed?

Comment (sonalkr132) [inline Gemfile]: It is used by https://github.com/rubygems/rubygems.org/blob/071df7d625ee18be64d225e78abb445fdbf17462/lib/tasks/gen_erd.rake

Comment (simi) [inline Gemfile]: Sure, but why changed?

Comment (sonalkr132) [inline Gemfile]: Oh! It has to do with rebase :stuck_out_tongue_closed_eyes: I had removed `quite_assets` changed group to one line but `quite_assets` was removed in different PR and I had to rebase.

I guess it saves two lines, but yes, the change is unnecessary. May be I can restore it when I update the PR for 5.0.1 if you think it is important.

Comment (simi) [inline Gemfile]: Since this PR is changing 169 files I think it will be nice to keep it as minimal as it is possible for future reviewers.

**Comment by simi** [comment]
I'm not sure if `ApplicationRecord` is needed and it will save a lot of line changes since it is not used (it is just blank abstract class). Is it really worth it now?

File: app/controllers/profiles_controller.rb [hunk 3]
```diff
@@ -5,10 +5,8 @@ def edit
   end
 
   def show
-    @user           = User.find_by_slug!(params[:id])
-    rubygems        = @user.rubygems_downloaded
-    @rubygems       = rubygems.slice!(0, 10)
-    @extra_rubygems = rubygems
```
Comment (simi) [inline app/controllers/profiles_controller.rb]: Is this removal of `@extragems_rubygems` and `rubygems` var assignments and following view change related to Rails 5 update or is it just a code cleanup/refactoring?

File: config/initializers/new_framework_defaults.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+# Make Ruby 2.4 preserve the timezone of the receiver when calling `to_time`.
+# Previous versions had false.
+ActiveSupport.to_time_preserves_timezone = false
+
+# Require `belongs_to` associations by default. Previous versions had false.
+Rails.application.config.active_record.belongs_to_required_by_default = false
+
+# Do not halt callback chains when a callback returns false. Previous versions had true.
+ActiveSupport.halt_callback_chains_on_return_false = false
```
Comment (simi) [inline config/initializers/new_framework_defaults.rb]: This is the only config option not respecting old versions and also resulting in code changes. Do we really need it right now?

**Comment by simi** [comment]
I left some notes @sonalkr132 to your awesome work!.
Anyway I'm just random passerby so it will be good to wait for someone from RubyGems.org team to reply if those comments make sense.

**Comment by dwradcliffe** [comment]
@simi has some good points. Smaller the changes the better.

**Comment by sonalkr132** [comment]
I have changed only `halt_callback_chains_on_return_false` config option because it was showing deprecation in test runs. That really irks me :persevere: I guess we can change it later.

I have removed commit which added rails version to migration as well because we don't *really* need that either. Migration deprecation show up only when migration are run..

**Comment by indirect** [comment]
@sonalkr132 can you re base this and update to 5.0.1 to fix the connection pool issue?

**Comment by sonalkr132** [comment]
Rails 5.0.1 is not released yet. [rails 5.0.1 Milestone](https://github.com/rails/rails/milestone/53)
Rails 5.0.0.1 was released and it was just a security update.

**Comment by indirect** [comment]
@sonalkr132 ah, I see. 😞 I guess this is blocked until Rails 5.0.1 comes out, then.

**Comment by arthurnn** [comment]
I agree 100% with @dwradcliffe . We cannot ship one big update PR to upgrade Rails version. it is too risky.
We need to slowly introduce the changes in master, and make sure they run on both versions. Once we have master able to run on rails 4.2 and 5, and we test it out the move to 5

File: test/functional/api/v1/profiles_controller_test.rb [hunk 5]
```diff
@@ -47,7 +47,7 @@ def response_body
       context "on GET to show when hide email" do
         setup do
           @user.update(hide_email: true)
-          get :show, id: @user.handle, format: format
+          get :show, params: { id: @user.handle }, format: format
```
Comment (connorshea) [inline test/functional/api/v1/profiles_controller_test.rb]: Changes like this should be done in a follow-up PR, in my opinion. This adds unnecessary review overhead to the Pull Request.

**Comment by connorshea** [comment]
This would be _a lot_ smaller if we didn't bother fixing deprecation warnings in this PR. We should focus on doing that afterward.

File: Gemfile [hunk 6]
```diff
@@ -16,7 +16,7 @@ gem 'daemons'
 gem 'dalli'
 gem 'delayed_job'
 gem 'delayed_job_active_record'
-gem 'doorkeeper'
+gem 'doorkeeper', '~> 4.0.0.rc4'
```
Comment (connorshea) [inline Gemfile]: This can be backported (and can use the stable 4.0.0 version)

File: Gemfile.lock [hunk 7]
```diff
@@ -176,50 +179,50 @@ GEM
...
     newrelic_rpm (3.16.0.318)
+    nio4r (1.2.1)
     nokogiri (1.6.8)
       mini_portile2 (~> 2.1.0)
       pkg-config (~> 1.1.7)
     parser (2.3.1.2)
       ast (~> 2.2)
-    paul_revere (2.0.0)
-      rails (~> 4.0)
+    paul_revere (2.1.0.rc1)
```
Comment (connorshea) [inline Gemfile.lock]: paul_revere seems to be [abandoned](https://github.com/thoughtbot/paul_revere/issues/35) from what I can tell, can we replace it with something else?

File: app/controllers/api/v1/api_keys_controller.rb [hunk 8]
```diff
@@ -7,9 +7,9 @@ def show
       sign_in User.authenticate(username, password)
       if current_user
         respond_to do |format|
-          format.any(:all) { render text: current_user.api_key }
+          format.any(:all) { render plain: current_user.api_key }
```
Comment (connorshea) [inline app/controllers/api/v1/api_keys_controller.rb]: Replacing `render text` with `render plain` seems like an unnecessary deprecation warning fix? Should be a follow-up PR.

File: config/initializers/filter_parameter_logging.rb [hunk 9]
```diff
@@ -0,0 +1,4 @@
+# Be sure to restart your server when you modify this file.
+
+# Configure sensitive parameters which will be filtered from the log file.
+Rails.application.config.filter_parameters += [:password]
```
Comment (connorshea) [inline config/initializers/filter_parameter_logging.rb]: Does this override an existing `filter_parameters` config?

File: app/controllers/api/base_controller.rb [hunk 10]
```diff
@@ -1,7 +1,6 @@
 class Api::BaseController < ApplicationController
   before_action :doorkeeper_authorize!, if: :doorkeeper_token
   before_action :authenticate_with_oauth, if: :doorkeeper_token
-  skip_before_action :require_ssl
```
Comment (connorshea) [inline app/controllers/api/base_controller.rb]: What's the explanation for this change?

File: Gemfile.lock [hunk 11]
```diff
@@ -142,7 +144,7 @@ GEM
       activesupport (>= 4.1.0)
     gravtastic (3.2.6)
     hashie (3.4.4)
-    high_voltage (2.4.0)
+    high_voltage (3.0.0)
```
Comment (connorshea) [inline Gemfile.lock]: This seems to be something we can backport.

File: Gemfile (hunk 6 above)
Comment (dwradcliffe) [inline Gemfile]: This is actually gone in master now.

**Comment by sonalkr132** [comment]
Thanks for review @connorshea I will keep your comments in mind when I make changes.

> We cannot ship one big update PR to upgrade Rails version. it is too risky.

If it's okay I would like to open a new PR with :point_up:. It doesn't it has to wait for rails 5.0.1 for people to review it.

> We need to slowly introduce the changes in master, and make sure they run on both versions.

I haven't read (blog/implementation) anything related to this. Can someone please explain it with a little more detail? Would it be something like [this](https://github.com/rubygems/rubygems.org/pull/1262/files#diff-2812c7f41474174f2a338265b679244aR86), where we check for rails versions?

**Comment by connorshea** [comment]
@sonalkr http://shayfrendt.com/posts/upgrading-github-to-rails-3-with-zero-downtime/

**Comment by connorshea** [comment]
See also what I'm doing for GitLab: https://gitlab.com/gitlab-org/gitlab-ce/issues/14286

Lots of things can be backported to master to make this PR a lot smaller.

**Comment by sonalkr132** [comment]
@connorshea thanks :) That was helpful. I will send a new PR soon.

**Comment by sonalkr132** [comment]
See #1615

**Pull request closed without merging**